### DISCUSSION TITLE
Axios error failing with 500 status code

### DISCUSSION THREAD

**Discussion by kelseypedersen** [vercel/next.js#38667] [Help]
Hi 👋 I have a NextJs app powered by a Rails backend. I'm stumped with an issue related to my Axios request from my NextJs app hosted on Vercel. 

- Make the `add something` request
- The Rails backend is creating the new record and returning it
- Yet - the network response status is 500 and the response is nil
- No idea what's happening - seems like it's getting blocked in the browser? 

Has anyone faced this before? Please any suggestions are welcome. 

Here's my current setup: 

**index.jsx**
```
addSomething(data)
  .then((res) => {
    console.log('then res', res)
    ... more things... 
  })
  .catch(({ error }) => {
    console.log('then error', error.response)
    ... more things... 
  })
```

**requests.js**
```
const addSomething = (_data) => {
  const response = axios({
    method: "POST",
    url: "/api/add-something",
    data: _data,
    headers: { "Content-Type": "multipart/form-data" },
  })

  return response;
};
```

**add-something.js**
```
export default async (req, res) => {
  const _data = JSON.parse(req.body);
  var formData = new FormData();
  formData.append("example", _data.example);
  const response = await fetch(`https://api.ourdomain.com/somethings`, {
    method: "POST",
    body: formData,
  })
  const data = await response.json();
  res.status(200).json(data);
})
```

**next.config.js**
```
module.exports = {
  async headers() {
    return [
      {
        // matching all API routes
        source: "/api/:path*",
        headers: [
          { key: "Access-Control-Allow-Credentials", value: "true" },
          { key: "Access-Control-Allow-Origin", value: "*" },
          { key: "Access-Control-Allow-Methods", value: "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
          { key: "Access-Control-Allow-Headers", value: "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" },
        ]
      }
    ]
  },
};
```

**Comment by Babbili** [comment]
@kelseypedersen  in `requests.js` it's form-data while in `add-something.js` it's json data, it's better to make them both json data
Try this in `requests.js`
```javascript
const addSomething = (_data) => {
  const response = axios.post('/api/add-something', {
    data: _data,    // try also data: JSON.stringify(_data) 
  })
  console.log('response', response)
  console.log('data', response.data)

  return response;
};
```

**Comment by icyJoseph** [comment]
Hi,

First, since `add-something` is running in an API route, then one problem is that `FormData` is not part of the `Node.js` API, only part of the browser spec.

So you can't really construct `var formData = new FormData();` on a Node.js environment. 

> Assuming you've installed a package or have created `FormData` in Node.js.

Second, API routes automatically serialize data to JSON, to disable this, export this config from the API route:

```js
export const config = {
  api: {
    bodyParser: false,
  },
}
```

The body will now be raw.

> A side note, since `_data` passed to Axios is probably form data already, you can skip specifying the header, https://github.com/axios/axios/issues/318#issuecomment-218948420, Axios is smart enough to know what content type to use.

Now we have to parse the raw body. I often use `formidable` for this:

```ts
import {Writable} from 'stream';

import formidable from 'formidable';
import {NextApiRequest, NextApiResponse, PageConfig} from 'next';

const formidableConfig = {
    keepExtensions: true,
    maxFileSize: 10_000_000,
    maxFieldsSize: 10_000_000,
    maxFields: 7,
    allowEmptyFiles: false,
    multiples: false,
};

function formidablePromise(
    req: NextApiRequest,
    opts?: Parameters<typeof formidable>[0]
): Promise<{fields: formidable.Fields; files: formidable.Files}> {
    return new Promise((accept, reject) => {
        const form = formidable(opts);

        form.parse(req, (err, fields, files) => {
            if (err) {
                return reject(err);
            }
            return accept({fields, files});
        });
    });
}

const fileConsumer = <T = unknown>(acc: T[]) => {
    const writable = new Writable({
        write: (chunk, _enc, next) => {
            acc.push(chunk);
            next();
        },
    });

    return writable;
};

export async function handler(req: NextApiRequest, res: NextApiResponse) {
    if (req.method !== 'POST') return res.status(404).end();

    try {
        const chunks: never[] = [];

        const {fields, files} = await formidablePromise(req, {
            ...formidableConfig,
            // consume this, otherwise formidable tries to save the file to disk
            fileWriteStreamHandler: () => fileConsumer(chunks),
        });

        console.log(fields);
        console.log(files);
        const {file} = files;

        const fileData = Buffer.concat(chunks); // or is it from? I always mix these up

        return res.status(204).end();
    } catch (err) {
        return res.status(500).json({error: 'Internal Server Error'});
    }
}

export const config: PageConfig = {
    api: {
        bodyParser: false,
    },
};

export default handler;
```

### Summary

1. Sending FormData, from to API with Axios or Fetch, doesn't need the header to be specified
2. The API handler needs to have its body parser disabled
3. Use a form data parser, such as formidable

In some cases it might be possible to handshake on JSON end to end, but to pass file data, that's not really possible, unless, we first upload the file to a CDN, and use a URL pointing to that CDN as our data. 

Do mind that there's also limits on the payload that API routes can handle, https://vercel.com/docs/concepts/limits/overview#serverless-function-payload-size-limit.
